### PR TITLE
feature: dynamic socket rooms & dynamic player screen

### PR DESCRIPTION
**Pull request by lkog77**
Contains sample logic for the allocation of rooms to each pair of players. Needs to be integrated and tested with the current duel logic. Also needs to be integrated + tested with tournament manager for allocation of bots.

Player screen dynamically renders winning component and choose your move component.

Also added a few other minor things:
- WaitingToStart component reads name from URL param.
- Added Allocate Players button in Host Screen for testing

### CODE REVIEW

File: client/src/pages/GameLobby.tsx [hunk 1]
```diff
@@ -13,6 +13,10 @@ const GameLobby = () => {
   const updateList = (player: PlayerAttributes) => {
     setPlayers((previousPlayers) => [...previousPlayers, player]);
   };
+  const handleAllocatePlayers = () => {
```
Comment (RoryT-U) [inline client/src/pages/GameLobby.tsx:16]: Should this be start game or is that different to allocate players?

File: client/src/pages/PlayerScreen.tsx [hunk 2]
```diff
@@ -1,24 +1,50 @@
...
 import WaitingToStart from "../components/WaitingToStart";
-//import ChoosePlayerMove from "../components/ChoosePlayerMove";
-//import CountDownTimer from "../components/CountDownTimer";
+import { socket } from "../App";
+import ChoosePlayerMove from "../components/ChoosePlayerMove";
+import CountDownTimer from "../components/CountDownTimer";
+import WinnerPlayer from "../components/WinnerPlayer";
 
 const PlayerScreen = () => {
+  const [shouldRenderComponents, setShouldRenderComponents] = useState(false);
```
Comment (RoryT-U) [inline client/src/pages/PlayerScreen.tsx:10]: probably should be a bit more descriptive, like what components are being rendered? Also could be shortened, to maybe just renderComponents or something, but that doesnt really matter.

File: server/server.ts [hunk 3]
```diff
@@ -96,6 +97,16 @@ io.on("connection", (socket) => {
     // Send GAME_CREATED event
     io.to(host.socketId).emit("GAME_CREATED", gameCode);
   });
+
+  socket.on("ALLOCATE_PLAYERS", async (gameCode) => {
+    const game: Game | undefined = gamesMap.get(gameCode);
+    if (!game) {
+      return;
+    }
+    const players = game.getPlayers();
```
Comment (RoryT-U) [inline server/server.ts:106]: Maybe make it so handleRoomAllocation can be called from the game class. Would rather not expose information when possible

File: server/socketRoomManager.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    if (winningSocket) {
+      // Emit a socket message to the winning player
+      winningSocket.emit("GAME_WINNER", { winner: winningPlayer });
+    }
+    return;
+  }
+};
+
+// Generate a random alphanumeric room name
+function generateUniqueRoomName(): string {
```
Comment (RoryT-U) [inline server/socketRoomManager.ts:63]: Rather then generating a random unique room name (which this isnt technically as you could potentially get the same random number eventually right?) you could instead just do some like "GAME_${gameCode}&ROOM_${roomNumber}".

**Review by RoryT-U — CHANGES_REQUESTED** [review]
Looks pretty good, just some small changes. Also, I was thinking that the PlayerScreen should incoporpate everything from waiting for host to start, to the duel ui until the duel has completely finished. This means that all these would be on the same page without any need to route between them. Lmk what you think of that. Could get slightly complicated and messy, which would be the main downside.

**Comment by RoryT-U** [comment]
Also may want to make some directories in the components for your components, if you have alot, as the components directory  is getting a bit big. But that obviously isn't just for you todo.

**Comment by lkog77** [comment]
> Looks pretty good, just some small changes. Also, I was thinking that the PlayerScreen should incoporpate everything from waiting for host to start, to the duel ui until the duel has completely finished. This means that all these would be on the same page without any need to route between them. Lmk what you think of that. Could get slightly complicated and messy, which would be the main downside.

Yeah i agree, was under the impression that this was going to be the case anyways.

**Review by RoryT-U — APPROVED** [review]
Yep looks all good now. I made a small change to the generation of room names, have a quick look over that when you can and then merge into dev.

**Pull request merged**